tenders/common/elf.c

```diff
@@ -73,120 +74,155 @@ static ssize_t pread_in_full(int fd, void *buf, size_t count, off_t offset)
-     * Load all segments with the LOAD directive from the elf file at offset
-     * p_offset, and copy that into p_addr in memory. The amount of bytes
-     * copied is p_filesz.  However, each segment should be given
-     * p_memsz aligned up to p_align bytes on memory.
+     * Load all program segments with the PT_LOAD directive.
      */
-    for (ph_i = 0; ph_i < ph_cnt; ph_i++) {
-        uint8_t *daddr;
-        uint64_t _end;
-        size_t offset = phdr[ph_i].p_offset;
-        size_t filesz = phdr[ph_i].p_filesz;
-        size_t memsz = phdr[ph_i].p_memsz;
-        uint64_t paddr = phdr[ph_i].p_paddr;
-        uint64_t align = phdr[ph_i].p_align;
-        uint64_t result;
-        int prot;
+    e_end = 0;
+    for (Elf64_Half ph_i = 0; ph_i < ehdr->e_phnum; ph_i++) {
+        Elf64_Addr p_vaddr = phdr[ph_i].p_vaddr;
+        Elf64_Xword p_filesz = phdr[ph_i].p_filesz;
+        Elf64_Xword p_memsz = phdr[ph_i].p_memsz;
+        Elf64_Xword p_align = phdr[ph_i].p_align;
+        Elf64_Addr p_vaddr_end;
 
         if (phdr[ph_i].p_type != PT_LOAD)
             continue;
 
-        if ((paddr >= mem_size) || add_overflow(paddr, filesz, result)
-                || (result >= mem_size))
+        if ((p_vaddr >= mem_size)
+                || add_overflow(p_vaddr, p_filesz, p_vaddr_end)
+                || (p_vaddr_end >= mem_size))
             goto out_invalid;
-        if (add_overflow(paddr, memsz, result) || (result >= mem_size))
+        /* 
+         * Each segment should be given p_memsz aligned up to p_align bytes of
+         * memory, compute this as p_vaddr_end.
+         */
+        if (add_overflow(p_vaddr, p_memsz, p_vaddr_end)
+                || (p_vaddr_end >= mem_size))
             goto out_invalid;
         /*
-         * Verify that align is a non-zero power of 2 and safely compute
-         * ((_end + (align - 1)) & -align).
+         * Verify that p_align is a non-zero power of 2 and safely compute
+         * ((p_vaddr_end + (p_align - 1)) & -p_align).
          */
-        if (align > 0 && (align & (align - 1)) == 0) {
-            if (add_overflow(result, (align - 1), _end))
+        if (p_align > 0 && (p_align & (p_align - 1)) == 0) {
+            if (add_overflow(p_vaddr_end, (p_align - 1), p_vaddr_end))
                 goto out_invalid;
-            _end = _end & -align;
-        }
-        else {
-            _end = result;
+            p_vaddr_end = p_vaddr_end & -p_align;
         }
-        if (_end > *p_end)
-            *p_end = _end;
+        /*
+         * Keep track of the highest byte of memory occupied by the program.
+         */
+        if (p_vaddr_end > e_end)
+            e_end = p_vaddr_end;
 
-        daddr = mem + paddr;
-        nbytes = pread_in_full(fd_kernel, daddr, filesz, offset);
+        /*
+         * Load the segment into host memory at host_vaddr and ensure any BSS
+         * (p_memsz - p_filesz) is initialised to zero.
+         */
+        uint8_t *host_vaddr = mem + p_vaddr;
```

Comment:
1) if `mem_size` is large (in turn permitting a large `p_vaddr`), and `mem` (allocated by the caller) is near the end of the address space, doesn't this risk overflowing?

2) Is it intentional that the start address is not adjusted for alignment, and that the only influence that `p_align` has is on `p_vaddr_end` (and in turn for the range for the `mprotect()` call)?

Reply:
> If mem_size is large (in turn permitting a large p_vaddr), and mem (allocated by the caller) is near the end of the address space, doesn't this risk overflowing?

I don't see how the caller could allocate a valid memory range (mem, memsize) where (mem + memsize) overlaps the address space. In any case, that would seem to be the caller's problem, not the ELF loader's?

> Is it intentional that the start address is not adjusted for alignment, and that the only influence that p_align has is on p_vaddr_end (and in turn for the range for the mprotect() call)?

No idea, I should re-read that part of the ELF spec.

Reply:
1) Yes, that is the caller's problem. But `mem_size` comes from the user, and `mem` is allocated by `malloc`, so I wonder if we should have something to handle this in place somewhere? (I don't know if we do). It seems like a failure mode that would be worth either checking for here, or documenting in `elf.h`?

Reply:
(1) Done in abeca34 using `assert()`s since this is a caller error.